### ISSUE TITLE
Allowing this in super call.

### ISSUE THREAD

**Issue by hayesmg** [google/traceur-compiler#1930]
I am attempting to pass a function to the superclass and receiving the "this is not allowed before super()" error. Is it possible to have it not inspect _this_ when it is wrapped in a delegate?

```js
class Bar extends Foo {
  constructor() {
    super(() => this.doSomething());
  }

  doSomething() { }
}
```

**Comment by caitp** [comment]
this should be doable:

- rewrite references to `this` within a class constructor declaration scope as a temp var
- holechecking in scopes that need it by testing if temp `this` variable is undefined

should work just fine, maybe

**Comment by hayesmg** [comment]
I've also tested it without the lambda expecting the delegate to be called using fn.call(...) and this also raises the error:

```js
class Bar extends Foo {
  constructor() {
    super(function() {
      // Expecting `this` to come from superclass call of fn.call(this, ...args).
      this.doSomething();
    });
  }

  doSomething() { }
}
```

**Comment by arv** [comment]
Right now we only do static checking for `this` in constructors. We would
have to combine that with a runtime check in these kind of situations.

On Sun, May 10, 2015 at 9:13 AM, hayesmg <notifications@github.com> wrote:

> I've also tested it without the lambda expecting the delegate to be called
> using fn.call(...) and this also raises the error:
>
> class Bar extends Foo {
>   constructor() {
>     super(function() {
>       // Expecting `this` to come from superclass call of fn.call(this, ...args).
>       this.doSomething();
>     });
>   }
>
>   doSomething() { }
> }
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/traceur-compiler/issues/1930#issuecomment-100639684>
> .
>



-- 
erik

**Comment by hayesmg** [comment]
@arv Thanks for the reply.  I'm guessing this might be enough of an edge case that you guys won't consider implementing it?  I suppose I could use 0.0.85 or introduce a method that is called after super that initializes the superclass' internal state.
